forge/last-checker.rkt

```diff
@@ -69,18 +69,22 @@
      (check-and-output formula
                        node/formula/quantified
                        checker-hash
-                       (begin (for-each
-                                (lambda (decl)
+                       (begin (foldl
+                                (lambda (decl sofar)
                                   (let ([var (car decl)]
                                         [domain (cdr decl)])
                                     ; CHECK: shadowed variables
                                     (when (assoc var quantvars)
                                       (raise-syntax-error #f
                                                           (format "Shadowing of variable ~a detected. Check for something like \"some x: A | some x : B | ...\"." var)
                                                           (datum->syntax #f var (build-source-location-syntax (nodeinfo-loc info)))))
-                                    ; CHECK: recur into domain(s)
-                                    (checkExpression run-or-state domain quantvars checker-hash)))
+                                    ; CHECK: recur into domain(s), carrying decls from prior quantifiers
+                                    (define one-more-decl (cons (assocify decl) sofar))
+                                    (checkExpression run-or-state domain one-more-decl checker-hash)
+                                    one-more-decl))
+                                quantvars 
                                 decls)
```

Comment:
Use the `foldl` output to define `new-quantvars` (below) instead of throwing it away.